torch/lib/THPP/tensors/generic/THTensor.cpp

```diff
@@ -249,12 +295,251 @@ auto THTensor<real>::unsqueeze(const Tensor& src, int dimension) -> THTensor& {
+#if defined(TH_REAL_IS_DOUBLE) || defined(TH_REAL_IS_FLOAT)
+  THTensor_(qr)(tensor, const_tensor_cast(rr).tensor, const_tensor_cast(a).tensor);
+  return *this;
+#else
+  throw std::runtime_error("Lapack operations are available only for floating point tensors");
+#endif
+}
+
+template<>
+auto THTensor<real>::geqrf(const Tensor& rtau, const Tensor& a) -> THTensor& {
+#if defined(TH_REAL_IS_DOUBLE) || defined(TH_REAL_IS_FLOAT)
+  THTensor_(geqrf)(tensor, const_tensor_cast(rtau).tensor, const_tensor_cast(a).tensor);
+  return *this;
+#else
+  throw std::runtime_error("Lapack operations are available only for floating point tensors");
+#endif
+}
+
+template<>
+auto THTensor<real>::orgqr(const Tensor& a, const Tensor& tau) -> THTensor& {
+#if defined(TH_REAL_IS_DOUBLE) || defined(TH_REAL_IS_FLOAT)
+  THTensor_(orgqr)(tensor, const_tensor_cast(a).tensor, const_tensor_cast(tau).tensor);
+  return *this;
+#else
+  throw std::runtime_error("Lapack operations are available only for floating point tensors");
+#endif
+}
+
+template<>
+auto THTensor<real>::ormqr(const Tensor& a, const Tensor& tau, const Tensor& c,
+                           const char *side, const char *trans) -> THTensor& {
+#if defined(TH_REAL_IS_DOUBLE) || defined(TH_REAL_IS_FLOAT)
+  THTensor_(ormqr)(tensor, const_tensor_cast(a).tensor, const_tensor_cast(tau).tensor,
+                   const_tensor_cast(c).tensor, side, trans);
+  return *this;
+#else
+  throw std::runtime_error("Lapack operations are available only for floating point tensors");
+#endif
+}
+
+template<>
+auto THTensor<real>::pstrf(const Tensor& rpiv, const Tensor& a,
+                           const char *uplo, scalar_type tol) -> THTensor& {
+#if defined(TH_REAL_IS_DOUBLE) || defined(TH_REAL_IS_FLOAT)
+  THTensor_(pstrf)(tensor, const_int_cast(rpiv).tensor, const_tensor_cast(a).tensor, uplo, tol);
+  return *this;
+#else
+  throw std::runtime_error("Lapack operations are available only for floating point tensors");
+#endif
+}
+
+template<>
 auto THTensor<real>::fill(scalar_type value) -> THTensor& {
   THTensor_(fill)(tensor, value);
   return *this;
 }
 
 template<>
+auto THTensor<real>::maskedFill(const Tensor& mask, scalar_type value) -> THTensor& {
+  THTensor_(maskedFill)(tensor, const_byte_cast(mask).tensor, value);
+  return *this;
+}
+
+template<>
+auto THTensor<real>::maskedCopy(const Tensor& mask, const Tensor& src) -> THTensor& {
+  THTensor_(maskedCopy)(tensor, const_byte_cast(mask).tensor, const_tensor_cast(src).tensor);
+  return *this;
+}
+
+template<>
+auto THTensor<real>::maskedSelect(const Tensor& src, const Tensor& mask) -> THTensor& {
+  THTensor_(maskedSelect)(tensor, const_tensor_cast(src).tensor, const_byte_cast(mask).tensor);
+  return *this;
+}
+
+template<>
+ptrdiff_t THTensor<real>::nonzeroElems() const {
+  ptrdiff_t result = 0;
+#ifdef TH_REAL_IS_HALF
+#define IS_NONZERO(val) ((val.x & 0x7fff) != 0)
+#else
+#define IS_NONZERO(val) ((val)!=0)
+#endif
+  TH_TENSOR_APPLY(real, tensor,
```

Comment:
This is just a helper for `THDTensor_(nonzero)` - the number of nonzero values is being computed by the worker and sent to master for the resize (so as not to use `THDTensor_(data)`), see [here](https://github.com/apaszke/pytorch-dist/pull/63/files#diff-c0e7ff130c09526e4236f3b315eecf23R73)